packages/metrics/src/features/metrics/utils.ts

```diff
@@ -109,6 +109,37 @@ export const generateEmptyBirthKeyFigure = (
   }
 }
 
+function createEstimatesArray(
+  years: number[],
+  output: number[],
+  extension: fhir.Extension
+) {
+  const locationExtensions = JSON.parse(extension.valueString as string) as {
+    [key: string]: number
+  }[]
+  // Imagine that years searching: [2022,2023] & extensions is: [2017,2022]
+  // We have no data for 2023 so we want to use 2022 estimates
+  for (let i = 0; i < years.length; i++) {
+    let year = years[i]
+    let entryIndex = locationExtensions.findIndex((obj) =>
+      obj.hasOwnProperty(year.toString())
+    )
+    while (entryIndex === -1) {
+      // If no extensions entry for that year find immediately preceeding year's data and use it.
+      // Keep going through the while loop till you find a matching year
+      year--
+      entryIndex = locationExtensions.findIndex((obj) =>
+        obj.hasOwnProperty(year.toString())
+      )
+    }
```

Comment:
The while loop caught my attention as there might be some cases where it could lead to an infinite loop. At least if "years" array contains a 2016 and extensions `[2017,2022]`. Potential also to happen if the arrays are of different lengths. The risk is that someone makes a query in production that creates an infinite loop and the backend halts for everyone.

Would it work to first sort both arrays ascending 

```
years: [2022, 2023]
extensions: [2017, 2022]
``

then take the last index of extension years where the year was smaller or equal to the year being searched

```ts
years.map((year) => 
  extensions.findLast((extension) => year <= extension)
)
```

Reply:
@rikukissa I was trying to find the closest year available to the supplied values.  I managed to find a different way to do it without using a while loop